extension/goplugin/environment.go

```diff
@@ -65,234 +65,233 @@ type SchemaPrioritizedSchemaHandlers map[string]PrioritizedSchemaHandlers
+func (env *Environment) SchemaHandlers() EventSchemaPrioritizedSchemaHandlers {
+	return env.schemaHandlers
 }
 
 // Core returns an implementation to Core interface
-func (thisEnvironment *Environment) Core() goext.ICore {
-	return thisEnvironment.extCore
+func (env *Environment) Core() goext.ICore {
+	return env.extCore
 }
 
 // Logger returns an implementation to Logger interface
-func (thisEnvironment *Environment) Logger() goext.ILogger {
-	return thisEnvironment.extLogger
+func (env *Environment) Logger() goext.ILogger {
+	return env.extLogger
 }
 
 // Schemas returns an implementation to Schemas interface
-func (thisEnvironment *Environment) Schemas() goext.ISchemas {
-	return thisEnvironment.extSchemas
+func (env *Environment) Schemas() goext.ISchemas {
+	return env.extSchemas
 }
 
 // Sync returns an implementation to Sync interface
-func (thisEnvironment *Environment) Sync() goext.ISync {
-	return thisEnvironment.extSync
+func (env *Environment) Sync() goext.ISync {
+	return env.extSync
 }
 
 // Database returns an implementation to IDatabase interface
-func (thisEnvironment *Environment) Database() goext.IDatabase {
-	return thisEnvironment.extDatabase
+func (env *Environment) Database() goext.IDatabase {
+	return env.extDatabase
 }
 
-//bind sets environment bindings
-func (thisEnvironment *Environment) bind() {
-	thisEnvironment.extCore = NewCore(thisEnvironment)
-	thisEnvironment.extLogger = NewLogger(thisEnvironment)
-	thisEnvironment.extSchemas = NewSchemas(thisEnvironment)
-	thisEnvironment.extSync = NewSync(thisEnvironment)
-	thisEnvironment.extDatabase = NewDatabase(thisEnvironment)
+// Http returns an implementation to IHttp interface
+func (env *Environment) Http() goext.IHttp {
+	return &Http{}
+}
+
+func (env *Environment) SetDatabase(db gohan_db.DB) {
+	env.db = db
+	env.bindDatabase()
+}
+
+func (env *Environment) SetIdentityService(ident middleware.IdentityService) {
+	env.ident = ident
+}
+
+func (env *Environment) SetSync(sync gohan_sync.Sync) {
+	env.sync = sync
+	env.bindSync()
+}
+
+func (env *Environment) bindCore() {
+	env.extCore = NewCore(env)
+}
+
+func (env *Environment) bindLogger() {
+	env.extLogger = NewLogger(env)
+}
+
+func (env *Environment) bindSchemas() {
+	env.extSchemas = NewSchemas(env)
+}
+
+func (env *Environment) bindSync() {
+	env.extSync = NewSync(env)
+}
+
+func (env *Environment) bindDatabase() {
+	env.extDatabase = NewDatabase(env)
 }
 
 // Start starts already loaded environment
-func (thisEnvironment *Environment) Start() error {
+func (env *Environment) Start() error {
 	var err error
 
-	if thisEnvironment.source == "" {
-		panic("golang extension is not loaded")
+	if len(env.initFns) == 0 {
+		log.Debug("Go environment is empty: %s", env.name)
+		return nil
 	}
 
-	log.Debug("Starting golang environment: %s", thisEnvironment.source)
+	log.Debug("Starting go environment: %s", env.name)
+
+	// bind
+	env.bindCore()
+	env.bindLogger()
+	env.bindSchemas()
```

Comment:
It looks there is no need to bind here because a constructor did the same thing.

Reply:
Right, there is a redundancy. In fact, there shouldn't be any binding in constructor and in clone. I removed those calls. Core, logger and Schemas is only available once environment is started and not yet stopped.